In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
# create path
path = r'OneDrive/Desktop/Instacart basket Analysis 12-2024'

In [3]:
# install dataframe
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared data', 'ords_prods_merge_updated.pkl'))

In [4]:
ords_prods_merge.head()

order_id  user_id  order_count  order_dow  order_hour_of_day  \
0   2539329        1            1          2                  8   
1   2539329        1            1          2                  8   
2   2539329        1            1          2                  8   
3   2539329        1            1          2                  8   
4   2539329        1            1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  \
0                                     Soda        77              7   
1  Organic Unsweetened Vanilla Almond Milk        91             16   
2                      Original Beef Jerky        23             19   
3               Aged White Cheddar Popcorn        23             19   
4         XL Pick-A-Size Paper Towel Rolls        54             17   

      prices    price_range_loc     busiest_day         busiest_days  \
0   9.000000  Mid-range product  Regularly busy  Regularly busy Days   
1  12.500000  Mid-range product  Regularly busy  Regularly busy Days   
2   4.398438  low-range product  Regularly busy  Regularly busy Days   
3   4.699219  low-range product  Regularly busy  Regularly busy Days   
4   1.000000  low-range product  Regularly busy  Regularly busy Days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2        Average orders  
3        Average orders  
4        Average orders

In [5]:
ords_prods_merge.shape

(32404859, 17)

In [6]:
# Grouping by 'department_id' and 'order_count', and Calculating Mean using agg() Function for DF

ords_prods_merge.groupby('department_id').agg({'order_count' : ['mean']})

order_count
                     mean
department_id            
1               15.457838
2               17.277920
3               17.170395
4               17.811403
5               15.215751
6               16.439806
7               17.225802
8               15.340650
9               15.895474
10              20.197148
11              16.170638
12              15.887671
13              16.583536
14              16.773669
15              16.165037
16              17.665606
17              15.694469
18              19.310397
19              17.177343
20              16.473447
21              22.902379

Question 3

When comparing the mean from a subset to the mean of an entire dataframe, the subset will always have a smaller mean considering it is using a portion of the dataframe compared to the whole thing. Using the whole dataframe will reveal more frequent ordering patterns by the users across the different departments compared to a portion caused by using a subset.

Ceate a loyalty flag

In [8]:
# Aggregating Data with transform() Function
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_count'].transform(np.max)

C:\Users\aaron\AppData\Local\Temp\ipykernel_15692\1830274060.py:2: FutureWarning: The provided callable <function max at 0x000001B839C98360> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_count'].transform(np.max)


In [9]:
# checking the output
ords_prods_merge.head(10)

order_id  user_id  order_count  order_dow  order_hour_of_day  \
0   2539329        1            1          2                  8   
1   2539329        1            1          2                  8   
2   2539329        1            1          2                  8   
3   2539329        1            1          2                  8   
4   2539329        1            1          2                  8   
5   2398795        1            2          3                  7   
6   2398795        1            2          3                  7   
7   2398795        1            2          3                  7   
8   2398795        1            2          3                  7   
9   2398795        1            2          3                  7   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   
5                    15.0         196                  1          1   
6                    15.0       10258                  2          0   
7                    15.0       12427                  3          1   
8                    15.0       13176                  4          0   
9                    15.0       26088                  5          1   

                              product_name  aisle_id  department_id  \
0                                     Soda        77              7   
1  Organic Unsweetened Vanilla Almond Milk        91             16   
2                      Original Beef Jerky        23             19   
3               Aged White Cheddar Popcorn        23             19   
4         XL Pick-A-Size Paper Towel Rolls        54             17   
5                                     Soda        77              7   
6                               Pistachios       117             19   
7                      Original Beef Jerky        23             19   
8                   Bag of Organic Bananas        24              4   
9               Aged White Cheddar Popcorn        23             19   

      prices    price_range_loc     busiest_day         busiest_days  \
0   9.000000  Mid-range product  Regularly busy  Regularly busy Days   
1  12.500000  Mid-range product  Regularly busy  Regularly busy Days   
2   4.398438  low-range product  Regularly busy  Regularly busy Days   
3   4.699219  low-range product  Regularly busy  Regularly busy Days   
4   1.000000  low-range product  Regularly busy  Regularly busy Days   
5   9.000000  Mid-range product  Regularly busy         Slowest Days   
6   3.000000  low-range product  Regularly busy         Slowest Days   
7   4.398438  low-range product  Regularly busy         Slowest Days   
8  10.296875  Mid-range product  Regularly busy         Slowest Days   
9   4.699219  low-range product  Regularly busy         Slowest Days   

  busiest_period_of_day  max_order  
0        Average orders         10  
1        Average orders         10  
2        Average orders         10  
3        Average orders         10  
4        Average orders         10  
5        Average orders         10  
6        Average orders         10  
7        Average orders         10  
8        Average orders         10  
9        Average orders         10

In [10]:
# Creating a 'Loyalty' Label Flag using loc Function
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [11]:
# Examining Value Counts on 'loyalty_flag'
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [12]:
# checking updated column
ords_prods_merge.head(15)

order_id  user_id  order_count  order_dow  order_hour_of_day  \
0    2539329        1            1          2                  8   
1    2539329        1            1          2                  8   
2    2539329        1            1          2                  8   
3    2539329        1            1          2                  8   
4    2539329        1            1          2                  8   
5    2398795        1            2          3                  7   
6    2398795        1            2          3                  7   
7    2398795        1            2          3                  7   
8    2398795        1            2          3                  7   
9    2398795        1            2          3                  7   
10   2398795        1            2          3                  7   
11    473747        1            3          3                 12   
12    473747        1            3          3                 12   
13    473747        1            3          3                 12   
14    473747        1            3          3                 12   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                      NaN       14084                  2          0   
2                      NaN       12427                  3          0   
3                      NaN       26088                  4          0   
4                      NaN       26405                  5          0   
5                     15.0         196                  1          1   
6                     15.0       10258                  2          0   
7                     15.0       12427                  3          1   
8                     15.0       13176                  4          0   
9                     15.0       26088                  5          1   
10                    15.0       13032                  6          0   
11                    21.0         196                  1          1   
12                    21.0       12427                  2          1   
13                    21.0       10258                  3          1   
14                    21.0       25133                  4          0   

                               product_name  aisle_id  department_id  \
0                                      Soda        77              7   
1   Organic Unsweetened Vanilla Almond Milk        91             16   
2                       Original Beef Jerky        23             19   
3                Aged White Cheddar Popcorn        23             19   
4          XL Pick-A-Size Paper Towel Rolls        54             17   
5                                      Soda        77              7   
6                                Pistachios       117             19   
7                       Original Beef Jerky        23             19   
8                    Bag of Organic Bananas        24              4   
9                Aged White Cheddar Popcorn        23             19   
10                    Cinnamon Toast Crunch       121             14   
11                                     Soda        77              7   
12                      Original Beef Jerky        23             19   
13                               Pistachios       117             19   
14                    Organic String Cheese        21             16   

       prices    price_range_loc     busiest_day         busiest_days  \
0    9.000000  Mid-range product  Regularly busy  Regularly busy Days   
1   12.500000  Mid-range product  Regularly busy  Regularly busy Days   
2    4.398438  low-range product  Regularly busy  Regularly busy Days   
3    4.699219  low-range product  Regularly busy  Regularly busy Days   
4    1.000000  low-range product  Regularly busy  Regularly busy Days   
5    9.000000  Mid-range product  Regularly busy         Slowest Days   
6    3.000000  low-range product  Regularly busy         Slowest Days   
7    4.398438  low-range 

Analyzing spending habits of the 3 loyalty groups

In [13]:
# Determining Basic Statistics of Spending Habits by Loyalty Category
ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773341  1.0  25.0
New customer      7.800966  1.0  25.0
Regular customer  7.798022  1.0  25.0

Loyal customers tend to purchase lower-costing items, with an average price of 7.77. Regular customers fall in the middle, averaging 7.79. The minimum amount spent for each group is 1.0, but the price max for all three is 25 meaning that is the most expensive thing in this dataset.

creating a spending flag

In [15]:
# Aggregating Data with transform() Function
ords_prods_merge['avg_product_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\aaron\AppData\Local\Temp\ipykernel_15692\1371496578.py:2: FutureWarning: The provided callable <function mean at 0x000001B839C98D60> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['avg_product_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [16]:
# Checking Relevant Columns 
ords_prods_merge[['user_id', 'avg_product_price']].head(15)

user_id  avg_product_price
0         1           6.367535
1         1           6.367535
2         1           6.367535
3         1           6.367535
4         1           6.367535
5         1           6.367535
6         1           6.367535
7         1           6.367535
8         1           6.367535
9         1           6.367535
10        1           6.367535
11        1           6.367535
12        1           6.367535
13        1           6.367535
14        1           6.367535

In [17]:
# Creating a 'Spending' Label Flag using loc Function
ords_prods_merge.loc[ords_prods_merge['avg_product_price'] >= 10, 'spending_flag'] = 'High spender'
ords_prods_merge.loc[ords_prods_merge['avg_product_price'] < 10, 'spending_flag'] = 'Low spender'

In [18]:
# Examining Value Count on 'spending_flag' Column
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     32285409
High spender      119450
Name: count, dtype: int64

In [19]:
# checking new columns
ords_prods_merge[['user_id', 'avg_product_price', 'spending_flag']].head(15)

user_id  avg_product_price spending_flag
0         1           6.367535   Low spender
1         1           6.367535   Low spender
2         1           6.367535   Low spender
3         1           6.367535   Low spender
4         1           6.367535   Low spender
5         1           6.367535   Low spender
6         1           6.367535   Low spender
7         1           6.367535   Low spender
8         1           6.367535   Low spender
9         1           6.367535   Low spender
10        1           6.367535   Low spender
11        1           6.367535   Low spender
12        1           6.367535   Low spender
13        1           6.367535   Low spender
14        1           6.367535   Low spender

In [20]:
# Aggregating Data with transform() Function
ords_prods_merge['median_days_between_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\aaron\AppData\Local\Temp\ipykernel_15692\981590324.py:2: FutureWarning: The provided callable <function median at 0x000001B839DF79C0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['median_days_between_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [21]:
# Checking Relevant Columns 
ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders']].head(15)

user_id  days_since_prior_order  median_days_between_orders
0         1                     NaN                        20.5
1         1                     NaN                        20.5
2         1                     NaN                        20.5
3         1                     NaN                        20.5
4         1                     NaN                        20.5
5         1                    15.0                        20.5
6         1                    15.0                        20.5
7         1                    15.0                        20.5
8         1                    15.0                        20.5
9         1                    15.0                        20.5
10        1                    15.0                        20.5
11        1                    21.0                        20.5
12        1                    21.0                        20.5
13        1                    21.0                        20.5
14        1                    21.0                        20.5

In [23]:
# Creating a 'Order Frequency' Label Flag using loc Function
ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['median_days_between_orders'] <= 20) & (ords_prods_merge['median_days_between_orders'] > 10), 'order_frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [24]:
# Examining Value Counts on 'order_frequency_flag'
ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [25]:
# Checking Relevant Columns of Updated Dataframe
ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders', 'order_frequency_flag']].head(15)

user_id  days_since_prior_order  median_days_between_orders  \
0         1                     NaN                        20.5   
1         1                     NaN                        20.5   
2         1                     NaN                        20.5   
3         1                     NaN                        20.5   
4         1                     NaN                        20.5   
5         1                    15.0                        20.5   
6         1                    15.0                        20.5   
7         1                    15.0                        20.5   
8         1                    15.0                        20.5   
9         1                    15.0                        20.5   
10        1                    15.0                        20.5   
11        1                    21.0                        20.5   
12        1                    21.0                        20.5   
13        1                    21.0                        20.5   
14        1                    21.0                        20.5   

     order_frequency_flag  
0   Non-frequent customer  
1   Non-frequent customer  
2   Non-frequent customer  
3   Non-frequent customer  
4   Non-frequent customer  
5   Non-frequent customer  
6   Non-frequent customer  
7   Non-frequent customer  
8   Non-frequent customer  
9   Non-frequent customer  
10  Non-frequent customer  
11  Non-frequent customer  
12  Non-frequent customer  
13  Non-frequent customer  
14  Non-frequent customer

exporting data as a pickle

In [26]:
# Exporting the Dataframe
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared data', 'orders_products_merged_grouped.pkl'))